### ISSUE TITLE
detect if SPI bus has been initialized (IDFGH-7142)

### ISSUE THREAD

**Issue by morganrallen** [espressif/esp-idf#8745]
This is a followup on #8642 
If a method was provided to detect if the bus was already initialized a decision could be made on calling `spi_bus_free` before reinitializing. In my situation I have a driver that supports multiple devices that all might not be initialized at the same time. I need a way to see if the bus need initializing (for the first device) or not (for the rest of them).

**Describe the solution you'd like**
Some function call like `bool spi_bus_initialized(spi_host_device_t host_id, spi_bus_config_t *bus_config)` or `spi_bus_is_initialized` for better clarity.

**Comment by KaeLL** [comment]
I think that's a trivial change, and there may even be an argument for creating an esp_err_t define specifically for denoting that something's already been initialized.  
Question, though: are you trying to have each peripheral's driver do all SPI initialization completely independent from one another?

**Comment by rtborg** [comment]
I'd also appreciate such functionality - for the reason above, each peripheral driver does SPI initialization independent from the others.

**Comment by abobija** [comment]
Bump 👀

**Comment by MassimilianoMontagni** [comment]
+1

**Comment by MassimilianoMontagni** [comment]
I'm using idf v 4.4.2 and I don't find `spi_bus_get_attr` do I import some header or it is no longer available?

**Comment by morganrallen** [comment]
Also using ~4.4.2, you'll need to include `driver/spi_common_internal.h`

**Comment by MassimilianoMontagni** [comment]
Thanks a lot @morganrallen but I still have the issue reported [here](https://github.com/espressif/esp-idf/issues/8642)

**Comment by abobija** [comment]
Is this resolved? Is there a way to check if SPI is initialized or not?

**Comment by abobija** [comment]
@KaeLL Sorry for the ping, but I see you are mentioning this issue frequently. Do you have any updates on this?

**Comment by KaeLL** [comment]
No. @igrr final word, maybe?

**Comment by morganrallen** [comment]
Something needs to get done, this is a needed feature, and my workaround doesn't even work anymore because `driver/spi_common_internal.h` isn't accessible anymore.

**Comment by igrr** [comment]
Reading the original issue description, I think two solutions are possible:

- SPI bus is initialized once at the application level, and SPI devices don't take care of initializing the bus. For example, if packaging device drivers as components, they would receive an SPI bus handle as an argument.
- If the choice of SPI devices to initialize happens at run time, and it's possible that no devices will be initialized, then we would probably not need to initialize the bus (since that would waste some RAM). In this case you can create a simple helper function for SPI initialization. It keeps a an initialization counter and only initializes the bus the first time it is called, A similar function for deinitializing the bus would first check the counter and free the bus when the counter reaches zero.

Both solutions above can be thread safe — either by construction (the first one) or by adding a mutex (the second one).

The approach with the getter function for SPI bus initialization status may, on the other hand, lead to race conditions. Imagine two tasks trying to use this "initialize if not yet initialized" pattern at the same time.

**Comment by KaeLL** [comment]
@igrr 
Yes, but take thread-safety off, and the solutions hide the inconsistent ways in which IDF approaches this problem. You get different results depending on which component you're dealing with. For instance, `esp_wifi_init` is idempotent, but `esp_netif_init`, `esp_event_loop_create_default`, among others, aren't. Also, this last group usually returns `ESP_ERR_INVALID_STATE`, which makes no sense because either there are side-effects and it's a misnomer warranting a new `esp_err_t` error code, or there aren't any side effects and it should just return `ESP_OK` like `esp_wifi_init`.

**Comment by ginkgm** [comment]
Hi @KaeLL ,

Here are some generic rules we follow in the design of driver APIs:

1. If the API involves resource allocation, we won't make it idempotent. And the return value will be explicit (Invalid state)
2. If the API involves some parameters in the configuration, we won't either.
3. If the peripheral itself can be a service and the API name looks like idempotent, we will make it idempotent, and the return value should be OK if repeated. For example, `enable` and `disable` is something I think should be idempotent.

Moreover, for performance reasons, we are not implementing locks for a same object (handle), instead, if users really want different tasks to work with a same object, they should implement the lock on their own.

But this may not apply to other areas due to historical and organization reasons.

For this specific case, bus initialization is obviously of class 1 and 2, so we won't make it idempotent. It makes sense for user to know the bus is initialized explicitly and their change in configuration value may fail. If you want to ensure the bus is initialized, it's better to have a common initialize interface to do this to:


1. Make sure the configuration used by all tasks are the same
2. You may even add a reference counter to make sure one task won't deinitialize the bus while there are other devices attached. 
3. Avoid possible issue when concurrent access to initialize API.
